### DISCUSSION TITLE
where can I find the code for nanoflownet?

### DISCUSSION THREAD

**Discussion by noahzn** [bitcraze/discussions#406] [Q&A] [answered]
Hi, I checked the iros2022 branch, but I didn't find anything related to nanoflownet. Could you tell me where I can find files related to the network? And, how should I run it? Thanks.

**Comment by ataffanel** [comment]
I am not sure if the source code has been published, it is as far as I know in none of the Bitcraze branches and repos (we did not use nanoflownet in any of our iros2022 demos).

@gemenerik should have more information as he is the author of the code.

**Comment by gemenerik** [comment, marked as answer]
Hi Noah. The source code has not yet been published. If you want to reproduce the results right now, I would recommend reading the [preprint of the article](https://arxiv.org/abs/2209.06918) and implementing the CNN in TensorFlow. For deploying the resulting CNN on the AI-deck, I would recommend watching [Greenwaves Technologies' BAM days talk](https://youtu.be/12s16Ex11EU) as a start. The implemented control algorithm is simple to replicate.

**Reply by noahzn** [reply]
Hi gemenerik, thanks for your reply. I remember yesterday that nanoflownet blog said the code was in the iros2022 branch, but today I cannot find this sentence anymore.  Can I implement networks in PyTorch? Or, now the tools only support TF?

**Reply by gemenerik** [reply]
I think that must have been another blogpost, this work is not IROS2022 related. 

[NNTOOL](https://github.com/GreenWaves-Technologies/gap_sdk/tree/master/tools/nntool) supports TensorFlow Lite or ONNX as input network. So yes, you can use PyTorch, and convert to ONNX. You would have to [verify the ONNX version supports all required operators](https://github.com/GreenWaves-Technologies/gap_sdk/blob/master/tools/nntool/README.md#operator-support) (conv2d, upsampling, etc). I think it does. 

Note: If you want to investigate quantization (e.g., quantization aware training) you can not currently input a quantized ONNX model into NNTOOL. If you are not interested in that, however, you can let NNTOOL perform the quantization as I did in the paper results.

**Reply by noahzn** [reply]
Thank you for the information!  I will start working on this.

**Reply by noahzn** [reply]
@gemenerik  Hi, may I know if your network outputs an optical flow map or a command? If you output the map, did you include the controlling implementations in the ONNX model? or implement them somewhere else? Thanks.

**Reply by gemenerik** [reply]
The network outputs an optical flow map. I implemented the balance strategy on the AI-deck and sent the resulting yaw rate commands to the Crazyflie over UART2.

**Comment by boomer319** [comment]
@gemenerik has the code been published yet?

I was planning on using NanoFlowNet on the AI deck for obstacle avoidance as part of my bachelor's thesis.
The difference to the implementation in the blog being: I want the crazyfly to modify the path I give it in the beginning when encountering an obstacle, also changing the other crazyflies trajectory in order not to take the others down.

Best regards

**Reply by gemenerik** [reply]
It has been published at https://github.com/tudelft/nanoflownet

**Reply by boomer319** [reply]
Thank you very much!
